modules/siddhi-extensions/output-mappers/xml-output-mapper/src/main/java/org/wso2/siddhi/extension/output/mapper/xml/XMLOutputMapper.java

```diff
@@ -20,53 +20,132 @@
+        if (validateXmlOption != null) {
+            xmlValidationEnabled = Boolean.parseBoolean(validateXmlOption.getValue());
+        }
     }
 
     /**
-     * Convert the given Event mapping to XML string
+     * Map and publish the given {@link Event} array
      *
-     * @param event         Event object
-     * @param mappedPayload mapped Payload if any
-     * @return the mapped XML string
+     * @param events                  Event object array
+     * @param outputTransportCallback output transport callback
+     * @param optionHolder            option holder containing static and dynamic options
+     * @param payloadTemplateBuilder  Unmapped payload for reference
      */
     @Override
-    public Object mapEvent(Event event, String mappedPayload) {
-        if (mappedPayload != null) {
-            return mappedPayload;
+    public void mapAndSend(Event[] events, OutputTransportCallback outputTransportCallback,
+                           OptionHolder optionHolder, TemplateBuilder payloadTemplateBuilder)
+            throws ConnectionUnavailableException {
+        if (events.length < 1) {
+            return;
+        }
+        StringBuilder sb = new StringBuilder();
+        if (payloadTemplateBuilder != null) {   //custom mapping
+            if (prefixXml != null) {
+                sb.append(prefixXml);
+            }
+            for (Event event : events) {
+                sb.append(payloadTemplateBuilder.build(event));
+            }
+            if (suffixXml != null) {
+                sb.append(suffixXml);
+            }
+            if (xmlValidationEnabled) {
+                DocumentBuilderFactory factory = DocumentBuilderFactory.newInstance();
+                factory.setValidating(false);
+                factory.setNamespaceAware(true);
+
+                try {
+                    DocumentBuilder builder = factory.newDocumentBuilder();
+                    builder.parse(new ByteArrayInputStream(sb.toString().getBytes()));
+                } catch (ParserConfigurationException e) {
+                    log.error("Error occurred when trying to validate output XML event for well-formedness. " +
+                            "Dropping event: " + sb.toString());
+                    return;
+                } catch (SAXException e) {
+                    log.error("Parse error occurred when validating output XML event for well-formedness. " +
+                            "Reason: " + e.getMessage() + "Dropping event: " + sb.toString());
+                    return;
+                } catch (IOException e) {
+                    log.error("IO error occurred when validating output XML event for well-formedness. " +
+                            "Reason: " + e.getMessage() + "Dropping event: " + sb.toString());
+                    return;
+                }
+            }
         } else {
-            return constructDefaultMapping(event);
+            for (Event event : events) {
```

Comment:
Since `events` is an array, it will first have to be converted to an iterable. It is uncertain whether there'll be much of a performance improvement by converting this to a Java8 forEach.